{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2789105e-f847-4eb9-a441-0eee17b4e75e",
   "metadata": {},
   "source": [
    "# Alliance Genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ae8fa9f7-3be1-4feb-95a4-51e6fa3132c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Environment Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0a2a4565-d5ed-4ec9-9034-9ddcbf551c2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "from concurrent.futures import ThreadPoolExecutor\n",
    "import json\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import gzip\n",
    "import io\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "58e06d70-54d9-4af3-84b0-f9429457d825",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Setting up data directories\n",
    "tmp_data = os.path.join('./resources/tmp_data')\n",
    "processed_data = os.path.join('./resources/processed_data')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6cc30a4b-a296-4a6d-93ea-5ed4db2c83e6",
   "metadata": {},
   "source": [
    "Utility Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "18bf1909-b402-4df5-965b-54c546bcb3ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def scrape_page(url, page, limit):\n",
    "    \"\"\"\n",
    "    Scrape a single page of records from the API.\n",
    "\n",
    "    Args:\n",
    "        url (str): The URL of the API endpoint.\n",
    "        page (int): The page number to scrape.\n",
    "        limit (int): The number of records per page.\n",
    "\n",
    "    Returns:\n",
    "        list: A list of records from the specified page.\n",
    "    \"\"\"\n",
    "    params = {'page': page, 'limit': limit}\n",
    "    response = requests.get(url, params=params)\n",
    "    data = response.json()\n",
    "    return data['results']\n",
    "\n",
    "def scrape_all_records(url, limit=20, max_workers=10):\n",
    "    \"\"\"\n",
    "    Scrape all records from the API.\n",
    "\n",
    "    Args:\n",
    "        url (str): The URL of the API endpoint.\n",
    "        limit (int, optional): The number of records per page. Defaults to 20.\n",
    "        max_workers (int, optional): The maximum number of concurrent requests. Defaults to 10.\n",
    "\n",
    "    Returns:\n",
    "        list: A list of all records from the API.\n",
    "    \"\"\"\n",
    "    response = requests.get(url, params={'page': 1, 'limit': limit})\n",
    "    data = response.json()\n",
    "    total_pages = -(-data['total'] // limit)  # ceiling division\n",
    "\n",
    "    with ThreadPoolExecutor(max_workers=max_workers) as executor:\n",
    "        futures = [executor.submit(scrape_page, url, page, limit) for page in range(1, total_pages + 1)]\n",
    "        all_records = [record for future in futures for record in future.result()]\n",
    "\n",
    "    return all_records"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "541767ff-739b-487b-ba19-547cc48742c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_gzipped_json(uri):\n",
    "    # Send a GET request to the URI\n",
    "    response = requests.get(uri)\n",
    "\n",
    "    # Check if the response was successful\n",
    "    if response.status_code != 200:\n",
    "        raise Exception(f\"Failed to download file from {uri}\")\n",
    "\n",
    "    # Create a BytesIO object to hold the gzipped data\n",
    "    gzipped_data = io.BytesIO(response.content)\n",
    "\n",
    "    # Create a GzipFile object to decompress the data\n",
    "    with gzip.GzipFile(fileobj=gzipped_data, mode='rb') as f:\n",
    "        # Read the decompressed data into a string\n",
    "        json_data = f.read().decode('utf-8')\n",
    "\n",
    "    # Parse the JSON data into a Python object\n",
    "    data = json.loads(json_data)\n",
    "\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d022c9b0-17aa-4ec7-a732-cfcdb5fd9f16",
   "metadata": {},
   "outputs": [],
   "source": [
    "from enum import Enum, auto\n",
    "import requests\n",
    "import os\n",
    "import json\n",
    "from biobox_analytics.utils import download_gzipped_json, oxo_mapping\n",
    "import io\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "\n",
    "class AllianceDownloads(Enum):\n",
    "\n",
    "    DISEASE_HUMAN = \"https://fms.alliancegenome.org/download/DISEASE-ALLIANCE-JSON_HUMAN.json.gz\"\n",
    "    DISEASE_MOUSE = \"https://fms.alliancegenome.org/download/DISEASE-ALLIANCE-JSON_MGI.json.gz\"\n",
    "\n",
    "\n",
    "class AllianceGenomeAdapter:\n",
    "\n",
    "    def __init__(self):\n",
    "        self._api = \"https://www.alliancegenome.org/api/\"\n",
    "        self._disease_gene_associations = []\n",
    "        self._prepare_mappers()\n",
    "        self._download_disease_associations()\n",
    "\n",
    "    def _prepare_mappers(self):\n",
    "\n",
    "        \n",
    "        human = pd.read_table(\"https://www.alliancegenome.org/api/geneMap/ensembl?species=NCBITaxon:9606\", sep=\"\\t\")\n",
    "\n",
    "        mouse = pd.read_table(\"https://www.alliancegenome.org/api/geneMap/ensembl?species=NCBITaxon:10090\", sep=\"\\t\")\n",
    "\n",
    "        self._gene_map = pd.concat([human, mouse])\n",
    "\n",
    "\n",
    "    def _download_disease_associations(self):\n",
    "        human_results = download_gzipped_json(AllianceDownloads.DISEASE_HUMAN.value)\n",
    "        DISEASE_IDS = list(set([x.get('DOID') for x in human_results['data']]))\n",
    "\n",
    "        # self._doid_mapping = oxo_mapping(DISEASE_IDS, [\"EFO\", \"MONDO\"])\n",
    "        with open(os.path.join('./resources/mappings', 'doid_efo.mapping.json'), 'r') as f:\n",
    "            \n",
    "            self._doid_mapping = json.load(f)\n",
    "\n",
    "        for _id in tqdm(DISEASE_IDS, desc=\"Collecting DiseaseGene Annotations\"):\n",
    " \n",
    "            try:\n",
    "                harmonized_disease_id = self._doid_mapping.get(_id)[0]\n",
    "            except:\n",
    "                continue\n",
    "            \n",
    "            records = scrape_all_records(f\"https://www.alliancegenome.org/api/disease/{_id}/genes?filter.species=Homo sapiens\")\n",
    "            associations = []\n",
    "            for anno in records:\n",
    "                association = self._format_disease_gene_annotation(anno, harmonized_disease_id)\n",
    "                if association is not None:\n",
    "                    associations.append(association)\n",
    "            self._disease_gene_associations.extend(associations)\n",
    "\n",
    "    def _hgnc2ensembl(self, hgnc):\n",
    "        subset = self._gene_map[self._gene_map['Gene ID'] == hgnc]\n",
    "        return subset['Ensembl ID'].values[0]\n",
    "    \n",
    "    def _format_disease_gene_annotation(self, anno, disease_id, fmt=\"edge\"):\n",
    "\n",
    "        relation_str = anno.get('generatedRelationString').replace('_', ' ')\n",
    "        disease_qualifiers = anno.get('diseaseQualifiers', [])\n",
    "\n",
    "        if len(disease_qualifiers) > 1:\n",
    "            print(anno)\n",
    "            raise Exception(\"Disease Qualifiers should not be more than one in length\")\n",
    "            \n",
    "        if len(disease_qualifiers) == 1:\n",
    "            disease_qualifier_str = disease_qualifiers[0].replace('_', ' ')\n",
    "            relation_str = f'{relation_str} {disease_qualifier_str}'\n",
    "\n",
    "        uuid = anno.get(\"uniqueId\")\n",
    "\n",
    "        references = anno.get('pubmedPubModIDs')\n",
    "        evidenceCodes = anno.get('evidenceCodes', [])\n",
    "        evidence = [x.get('abbreviation') for x in evidenceCodes]\n",
    "        \n",
    "\n",
    "        try:\n",
    "            gene_hgnc = anno['subject']['modEntityId']\n",
    "            ensembl_id = self._hgnc2ensembl(gene_hgnc)\n",
    "        except:\n",
    "            return\n",
    "\n",
    "        output = {\n",
    "            'from': {\n",
    "                'uuid': ensembl_id\n",
    "            },\n",
    "            'to': {\n",
    "                'uuid': disease_id\n",
    "            },\n",
    "            'label': relation_str,\n",
    "            'properties': {\n",
    "                'uuid': uuid,\n",
    "                'references': references,\n",
    "                'evidence': evidence,\n",
    "                'citation_authority': 'Alliance of Genome Resources'\n",
    "            }\n",
    "        }\n",
    "        return output\n",
    "        \n",
    "\n",
    "    \n",
    "    # def get_nodes(self):\n",
    "    #     \"\"\"\n",
    "    #     Returns a generator of node tuples for node types specified in the\n",
    "    #     adapter constructor.\n",
    "    #     \"\"\"\n",
    "    #     print(\"Generating Nodes\")\n",
    "\n",
    "    #     for da in self._disease_gene_associations:\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8a8b2853-3891-4218-bc41-11e25fbb0ca6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Collecting DiseaseGene Annotations: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 6127/6127 [46:03<00:00,  2.22it/s]\n"
     ]
    }
   ],
   "source": [
    "adapter = AllianceGenomeAdapter()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f8f2df0c-b15c-41ad-9bf4-5e1606d7af6f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "234908"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(adapter._disease_gene_associations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1975e298-78ee-4581-867d-abff1c113bbd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'from': {'uuid': 'ENSG00000156170'},\n",
       " 'to': {'uuid': 'MONDO:0009723'},\n",
       " 'label': 'is implicated via orthology',\n",
       " 'properties': {'uuid': 'wBCpG48BJgBeDSVnk52w',\n",
       "  'references': ['MGI:6194238'],\n",
       "  'evidence': ['IEA'],\n",
       "  'citation_authority': 'Alliance of Genome Resources'}}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adapter._disease_gene_associations[100]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5ef5da71-b6d6-4d26-a201-941016c1acdd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 234908/234908 [00:03<00:00, 74724.49it/s]\n"
     ]
    }
   ],
   "source": [
    "with gzip.open(os.path.join('./resources/processed_data', 'alliance_disease_gene_edge.jsonl.gz'), 'wt') as f:\n",
    "    for n in tqdm(adapter._disease_gene_associations):\n",
    "        f.write(json.dumps(n) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c204a00e-19e1-438b-840b-e3cf6ee681eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "rel_labels = list(set([x.get('label') for x in adapter._disease_gene_associations]))\n",
    "relationships = {\n",
    "    x: {'from': 'Gene', 'to': 'Disease'} for x in rel_labels\n",
    "}\n",
    "\n",
    "alliance_gene_disease_associations_metadata = {\n",
    "  \"_meta\": {\n",
    "    \"version\": \"0.1.0\",\n",
    "    \"date_updated\": \"2024-05-20\",\n",
    "    \"maintainer\": \"BioBox Analytics\"\n",
    "  },\n",
    "  \"key\": \"alliance_disease_gene\",\n",
    "  \"name\": \"Alliance - Disease Gene Assocations\",\n",
    "  \"description\": \"A consortium of 7 model organism databases (MODs) and the Gene Ontology (GO) Consortium whose goal is to provide an integrated view of their data to all biologists, clinicians and other interested parties. This data package contains edges between Genes and Diseases. It will require you to have loaded in EFO and MONDO ontologies.\",\n",
    "  \"source\": [\n",
    "    {\n",
    "      \"uri\": \"https://www.alliancegenome.org/downloads\",\n",
    "      \"type\": \"doc\"\n",
    "    },\n",
    "    {\n",
    "      \"uri\": \"https://fms.alliancegenome.org/download/DISEASE-ALLIANCE-JSON_HUMAN.json.gz\",\n",
    "      \"type\": \"data\",\n",
    "      \"version\": \"7.1.0\"\n",
    "    }\n",
    "  ],\n",
    "  \"concepts\": {},\n",
    "  \"relationships\": relationships\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "129c8254-3628-40bd-b33a-a5016c027ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./resources/processed_data/alliance_disease_gene_metadata.json', 'w') as outfile:\n",
    "    json.dump(alliance_gene_disease_associations_metadata, outfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60c7c03d-9d86-4f8e-9355-5a4aedea87eb",
   "metadata": {},
   "source": [
    "# Scratchpad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "14a40461-1d37-45e9-9581-2e19ef2ad42d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>1</td>\n",
       "      <td>A1bg</td>\n",
       "      <td>MGI:2152878</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>29974</td>\n",
       "      <td>A1cf</td>\n",
       "      <td>MGI:1917115</td>\n",
       "      <td>MP:0005367, MP:0005369, MP:0005370, MP:0005376...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M</td>\n",
       "      <td>2</td>\n",
       "      <td>A2m</td>\n",
       "      <td>MGI:2449119</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>127550</td>\n",
       "      <td>A3galt2</td>\n",
       "      <td>MGI:2685279</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>53947</td>\n",
       "      <td>A4galt</td>\n",
       "      <td>MGI:3512453</td>\n",
       "      <td>MP:0005376, MP:0005386, MP:0010768</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         0       1        2            3  \\\n",
       "0     A1BG       1     A1bg  MGI:2152878   \n",
       "1     A1CF   29974     A1cf  MGI:1917115   \n",
       "2      A2M       2      A2m  MGI:2449119   \n",
       "3  A3GALT2  127550  A3galt2  MGI:2685279   \n",
       "4   A4GALT   53947   A4galt  MGI:3512453   \n",
       "\n",
       "                                                   4   5  \n",
       "0                                                NaN NaN  \n",
       "1  MP:0005367, MP:0005369, MP:0005370, MP:0005376... NaN  \n",
       "2                                                NaN NaN  \n",
       "3                                                NaN NaN  \n",
       "4                 MP:0005376, MP:0005386, MP:0010768 NaN  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hmd_human_phenotype.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "aa4e65f2-19dd-42c5-8236-e9f75854bb43",
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_tab_delimited_data(url, col_names):\n",
    "    \"\"\"Downloads tab-delimited data from a URL, skipping header lines and empty lines.\n",
    "\n",
    "    Args:\n",
    "        url: The URL of the tab-delimited file.\n",
    "\n",
    "    Returns:\n",
    "        A pandas DataFrame containing the data.\n",
    "    \"\"\"\n",
    "\n",
    "    response = requests.get(url)\n",
    "    response.raise_for_status()  # Raise an exception if the request failed\n",
    "\n",
    "    # Filter out header lines and empty lines\n",
    "    data_lines = [line for line in response.text.splitlines() \n",
    "                  if not line.startswith('#')]\n",
    "\n",
    "    # Create a file-like object from the data\n",
    "\n",
    "    data_io = io.StringIO('\\n'.join(data_lines)) \n",
    "\n",
    "    # Read the data into a DataFrame, specifying the delimiter and quoting\n",
    "    df = pd.read_table(data_io, sep='\\t', names=col_names, header=None, skip_blank_lines=True, on_bad_lines='skip')\n",
    "\n",
    "    return df\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f7de9ee-59ed-4085-9334-4dfd6d8de346",
   "metadata": {},
   "source": [
    "# MGI: Alleles and Genotypes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16812940-413b-412e-b49a-4ee6fd887b79",
   "metadata": {},
   "source": [
    "2 Alleles -> 1 Genotype - model of -> Disease"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "ed3540bc-633b-49a6-8ed8-08688451c9f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "mgi_disease_mouse_model = download_tab_delimited_data(\n",
    "    \"https://www.informatics.jax.org/downloads/reports/MGI_DiseaseMouseModel.rpt\",\n",
    "    ['do_term', 'DOID', 'negated', 'allele_pairs', 'strain', 'allele_symbol',\n",
    "     'allele_mgi_id', 'totalNumAlleleReferences', 'repo_id', 'mouse_genotype_rr_id', 'mouse_gene_symbol', 'mouse_gene_id', 'repo_id_gene'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "bd679b06-550a-4439-9e7e-1d046a234416",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>do_term</th>\n",
       "      <th>DOID</th>\n",
       "      <th>negated</th>\n",
       "      <th>allele_pairs</th>\n",
       "      <th>strain</th>\n",
       "      <th>allele_symbol</th>\n",
       "      <th>allele_mgi_id</th>\n",
       "      <th>totalNumAlleleReferences</th>\n",
       "      <th>repo_id</th>\n",
       "      <th>mouse_genotype_rr_id</th>\n",
       "      <th>mouse_gene_symbol</th>\n",
       "      <th>mouse_gene_id</th>\n",
       "      <th>repo_id_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>17-beta hydroxysteroid dehydrogenase 3 deficiency</td>\n",
       "      <td>DOID:0112248</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hsd17b3&lt;tm1.2Mpo&gt;/Hsd17b3&lt;tm1.2Mpo&gt;</td>\n",
       "      <td>involves: 129S6/SvEvTac * C57BL/6N * C57BL/6NCrl</td>\n",
       "      <td>Hsd17b3&lt;tm1.2Mpo&gt;</td>\n",
       "      <td>MGI:6508689</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RRID:MGI:6508693</td>\n",
       "      <td>Hsd17b3</td>\n",
       "      <td>MGI:107177</td>\n",
       "      <td>EuMMCR:12655|EuMMCR:3397|EuMMCR:3398|MMRRC:069...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2-aminoadipic 2-oxoadipic aciduria</td>\n",
       "      <td>DOID:0111453</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Dhtkd1&lt;em1Hpr&gt;/Dhtkd1&lt;em1Hpr&gt;</td>\n",
       "      <td>involves: C57BL/6N * FVB/N</td>\n",
       "      <td>Dhtkd1&lt;em1Hpr&gt;</td>\n",
       "      <td>MGI:6268324</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RRID:MGI:6268326</td>\n",
       "      <td>Dhtkd1</td>\n",
       "      <td>MGI:2445096</td>\n",
       "      <td>NM-KI-00005|MMRRC:042792-MU|MMRRC:045216-MU|MM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3-methylglutaconic aciduria type 3</td>\n",
       "      <td>DOID:0110004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Opa3&lt;m1Votr&gt;/Opa3&lt;m1Votr&gt;</td>\n",
       "      <td>involves: C3H * C57BL/6JCrl</td>\n",
       "      <td>Opa3&lt;m1Votr&gt;</td>\n",
       "      <td>MGI:5312669</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RRID:MGI:5312681</td>\n",
       "      <td>Opa3</td>\n",
       "      <td>MGI:2686271</td>\n",
       "      <td>EuMMCR:25483|MMRRC:067719-MU|MMRRC:041675-MU|M...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3-methylglutaconic aciduria with deafness, enc...</td>\n",
       "      <td>DOID:0110001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serac1&lt;em1Bcgen&gt;/Serac1&lt;em1Bcgen&gt;</td>\n",
       "      <td>C57BL/6N-Serac1&lt;em1Bcgen&gt;</td>\n",
       "      <td>Serac1&lt;em1Bcgen&gt;</td>\n",
       "      <td>MGI:7316653</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RRID:MGI:7316678</td>\n",
       "      <td>Serac1</td>\n",
       "      <td>MGI:2447813</td>\n",
       "      <td>CMMR:CMMR:ADPW|NM-KO-2100721|MMRRC:069450-MU|M...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>46 XX gonadal dysgenesis</td>\n",
       "      <td>DOID:14450</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fancl&lt;gcd&gt;/Fancl&lt;gcd&gt;</td>\n",
       "      <td>involves: C57BL/6J * CBA/J</td>\n",
       "      <td>Fancl&lt;gcd&gt;</td>\n",
       "      <td>MGI:1861640</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RRID:MGI:2660751</td>\n",
       "      <td>Fancl</td>\n",
       "      <td>MGI:1914280</td>\n",
       "      <td>EuMMCR:9783|EuMMCR:9782|MMRRC:044564-MU|MMRRC:...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             do_term          DOID negated  \\\n",
       "0  17-beta hydroxysteroid dehydrogenase 3 deficiency  DOID:0112248     NaN   \n",
       "1                 2-aminoadipic 2-oxoadipic aciduria  DOID:0111453     NaN   \n",
       "2                 3-methylglutaconic aciduria type 3  DOID:0110004     NaN   \n",
       "3  3-methylglutaconic aciduria with deafness, enc...  DOID:0110001     NaN   \n",
       "4                           46 XX gonadal dysgenesis    DOID:14450     NaN   \n",
       "\n",
       "                          allele_pairs  \\\n",
       "0  Hsd17b3<tm1.2Mpo>/Hsd17b3<tm1.2Mpo>   \n",
       "1        Dhtkd1<em1Hpr>/Dhtkd1<em1Hpr>   \n",
       "2            Opa3<m1Votr>/Opa3<m1Votr>   \n",
       "3    Serac1<em1Bcgen>/Serac1<em1Bcgen>   \n",
       "4                Fancl<gcd>/Fancl<gcd>   \n",
       "\n",
       "                                             strain      allele_symbol  \\\n",
       "0  involves: 129S6/SvEvTac * C57BL/6N * C57BL/6NCrl  Hsd17b3<tm1.2Mpo>   \n",
       "1                        involves: C57BL/6N * FVB/N     Dhtkd1<em1Hpr>   \n",
       "2                       involves: C3H * C57BL/6JCrl       Opa3<m1Votr>   \n",
       "3                         C57BL/6N-Serac1<em1Bcgen>   Serac1<em1Bcgen>   \n",
       "4                        involves: C57BL/6J * CBA/J         Fancl<gcd>   \n",
       "\n",
       "  allele_mgi_id  totalNumAlleleReferences repo_id mouse_genotype_rr_id  \\\n",
       "0   MGI:6508689                       1.0     NaN     RRID:MGI:6508693   \n",
       "1   MGI:6268324                       1.0     NaN     RRID:MGI:6268326   \n",
       "2   MGI:5312669                       5.0     NaN     RRID:MGI:5312681   \n",
       "3   MGI:7316653                       1.0     NaN     RRID:MGI:7316678   \n",
       "4   MGI:1861640                       5.0     NaN     RRID:MGI:2660751   \n",
       "\n",
       "  mouse_gene_symbol mouse_gene_id  \\\n",
       "0           Hsd17b3    MGI:107177   \n",
       "1            Dhtkd1   MGI:2445096   \n",
       "2              Opa3   MGI:2686271   \n",
       "3            Serac1   MGI:2447813   \n",
       "4             Fancl   MGI:1914280   \n",
       "\n",
       "                                        repo_id_gene  \n",
       "0  EuMMCR:12655|EuMMCR:3397|EuMMCR:3398|MMRRC:069...  \n",
       "1  NM-KI-00005|MMRRC:042792-MU|MMRRC:045216-MU|MM...  \n",
       "2  EuMMCR:25483|MMRRC:067719-MU|MMRRC:041675-MU|M...  \n",
       "3  CMMR:CMMR:ADPW|NM-KO-2100721|MMRRC:069450-MU|M...  \n",
       "4  EuMMCR:9783|EuMMCR:9782|MMRRC:044564-MU|MMRRC:...  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mgi_disease_mouse_model.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e14f497e-b5a8-47c8-b51e-1ad982eb1b86",
   "metadata": {},
   "source": [
    "Gather all the alleles as nodes first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "d87dbfc4-0e0c-48b8-b21c-455d2fed1ee3",
   "metadata": {},
   "outputs": [],
   "source": [
    "mgi_phenotypic_allele = download_tab_delimited_data('https://www.informatics.jax.org/downloads/reports/MGI_PhenotypicAllele.rpt',\n",
    "                                                   ['mgi_allele_id', 'allele_symbol', 'allele_name', 'allele_type', 'allele_attribute', 'PMID', 'mgi_marker_id','marker_symbol', 'marker_refseq_id', 'marker_ensembl_id', 'mp_id', 'synonyms', 'marker_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "bff31e92-97cb-45df-81a4-207453fa90e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mgi_allele_id</th>\n",
       "      <th>allele_symbol</th>\n",
       "      <th>allele_name</th>\n",
       "      <th>allele_type</th>\n",
       "      <th>allele_attribute</th>\n",
       "      <th>PMID</th>\n",
       "      <th>mgi_marker_id</th>\n",
       "      <th>marker_symbol</th>\n",
       "      <th>marker_refseq_id</th>\n",
       "      <th>marker_ensembl_id</th>\n",
       "      <th>mp_id</th>\n",
       "      <th>synonyms</th>\n",
       "      <th>marker_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGI:5603349</td>\n",
       "      <td>0610010K14Rik&lt;tm1.1(KOMP)Vlcg&gt;</td>\n",
       "      <td>targeted mutation 1.1, Velocigene</td>\n",
       "      <td>Targeted</td>\n",
       "      <td>Null/knockout|Reporter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGI:1915609</td>\n",
       "      <td>0610010K14Rik</td>\n",
       "      <td>NM_001177601</td>\n",
       "      <td>ENSMUSG00000020831</td>\n",
       "      <td>MP:0005386,MP:0010768</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RIKEN cDNA 0610010K14 gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MGI:4452997</td>\n",
       "      <td>0610010K14Rik&lt;tm1(KOMP)Vlcg&gt;</td>\n",
       "      <td>targeted mutation 1, Velocigene</td>\n",
       "      <td>Targeted</td>\n",
       "      <td>Null/knockout|Reporter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGI:1915609</td>\n",
       "      <td>0610010K14Rik</td>\n",
       "      <td>NM_001177601</td>\n",
       "      <td>ENSMUSG00000020831</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RIKEN cDNA 0610010K14 gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:7425474</td>\n",
       "      <td>0610040J01Rik&lt;em1(IMPC)Ccpcz&gt;</td>\n",
       "      <td>endonuclease-mediated mutation 1, Institute of...</td>\n",
       "      <td>Endonuclease-mediated</td>\n",
       "      <td>Null/knockout</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGI:1923511</td>\n",
       "      <td>0610040J01Rik</td>\n",
       "      <td>NM_001417897</td>\n",
       "      <td>ENSMUSG00000060512</td>\n",
       "      <td>MP:0005376,MP:0005378,MP:0005385,MP:0005387,MP...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RIKEN cDNA 0610040J01 gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MGI:3923426</td>\n",
       "      <td>1110038B12Rik&lt;Gt(Ayu21-B141)Imeg&gt;</td>\n",
       "      <td>gene trap Ayu21-B141, Institute of Molecular E...</td>\n",
       "      <td>Gene trapped</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGI:1916013</td>\n",
       "      <td>1110038B12Rik</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000092203</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RIKEN cDNA 1110038B12 gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MGI:3923883</td>\n",
       "      <td>1110059E24Rik&lt;Gt(Ayu21-W256)Imeg&gt;</td>\n",
       "      <td>gene trap Ayu21-W256, Institute of Molecular E...</td>\n",
       "      <td>Gene trapped</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGI:1913456</td>\n",
       "      <td>1110059E24Rik</td>\n",
       "      <td>NM_025423</td>\n",
       "      <td>ENSMUSG00000035171</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RIKEN cDNA 1110059E24 gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  mgi_allele_id                      allele_symbol  \\\n",
       "0   MGI:5603349     0610010K14Rik<tm1.1(KOMP)Vlcg>   \n",
       "1   MGI:4452997       0610010K14Rik<tm1(KOMP)Vlcg>   \n",
       "2   MGI:7425474      0610040J01Rik<em1(IMPC)Ccpcz>   \n",
       "3   MGI:3923426  1110038B12Rik<Gt(Ayu21-B141)Imeg>   \n",
       "4   MGI:3923883  1110059E24Rik<Gt(Ayu21-W256)Imeg>   \n",
       "\n",
       "                                         allele_name            allele_type  \\\n",
       "0                  targeted mutation 1.1, Velocigene               Targeted   \n",
       "1                    targeted mutation 1, Velocigene               Targeted   \n",
       "2  endonuclease-mediated mutation 1, Institute of...  Endonuclease-mediated   \n",
       "3  gene trap Ayu21-B141, Institute of Molecular E...           Gene trapped   \n",
       "4  gene trap Ayu21-W256, Institute of Molecular E...           Gene trapped   \n",
       "\n",
       "         allele_attribute  PMID mgi_marker_id  marker_symbol marker_refseq_id  \\\n",
       "0  Null/knockout|Reporter   NaN   MGI:1915609  0610010K14Rik     NM_001177601   \n",
       "1  Null/knockout|Reporter   NaN   MGI:1915609  0610010K14Rik     NM_001177601   \n",
       "2           Null/knockout   NaN   MGI:1923511  0610040J01Rik     NM_001417897   \n",
       "3                     NaN   NaN   MGI:1916013  1110038B12Rik              NaN   \n",
       "4                     NaN   NaN   MGI:1913456  1110059E24Rik        NM_025423   \n",
       "\n",
       "    marker_ensembl_id                                              mp_id  \\\n",
       "0  ENSMUSG00000020831                              MP:0005386,MP:0010768   \n",
       "1  ENSMUSG00000020831                                                NaN   \n",
       "2  ENSMUSG00000060512  MP:0005376,MP:0005378,MP:0005385,MP:0005387,MP...   \n",
       "3  ENSMUSG00000092203                                                NaN   \n",
       "4  ENSMUSG00000035171                                                NaN   \n",
       "\n",
       "  synonyms                 marker_name  \n",
       "0      NaN  RIKEN cDNA 0610010K14 gene  \n",
       "1      NaN  RIKEN cDNA 0610010K14 gene  \n",
       "2      NaN  RIKEN cDNA 0610040J01 gene  \n",
       "3      NaN  RIKEN cDNA 1110038B12 gene  \n",
       "4      NaN  RIKEN cDNA 1110059E24 gene  "
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mgi_phenotypic_allele.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "b3f61ada-23ff-448c-b61f-3eee854058ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mgi_allele_id</th>\n",
       "      <th>allele_symbol</th>\n",
       "      <th>allele_name</th>\n",
       "      <th>allele_type</th>\n",
       "      <th>allele_attribute</th>\n",
       "      <th>PMID</th>\n",
       "      <th>mgi_marker_id</th>\n",
       "      <th>marker_symbol</th>\n",
       "      <th>marker_refseq_id</th>\n",
       "      <th>marker_ensembl_id</th>\n",
       "      <th>mp_id</th>\n",
       "      <th>synonyms</th>\n",
       "      <th>marker_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>108940</td>\n",
       "      <td>108940</td>\n",
       "      <td>108937</td>\n",
       "      <td>108940</td>\n",
       "      <td>99813</td>\n",
       "      <td>5.049500e+04</td>\n",
       "      <td>108940</td>\n",
       "      <td>108940</td>\n",
       "      <td>89676</td>\n",
       "      <td>92229</td>\n",
       "      <td>44935</td>\n",
       "      <td>72449</td>\n",
       "      <td>108940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>108940</td>\n",
       "      <td>108867</td>\n",
       "      <td>42229</td>\n",
       "      <td>14</td>\n",
       "      <td>1048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>33118</td>\n",
       "      <td>33115</td>\n",
       "      <td>17298</td>\n",
       "      <td>17741</td>\n",
       "      <td>13572</td>\n",
       "      <td>70938</td>\n",
       "      <td>31283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>MGI:5603349</td>\n",
       "      <td>Arhgap6/Hccs/Mid1&lt;tm1Hzo&gt;</td>\n",
       "      <td>endonuclease-mediated mutation 1, GemPharmatec...</td>\n",
       "      <td>Targeted</td>\n",
       "      <td>Null/knockout</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGI:104735</td>\n",
       "      <td>Gt(ROSA)26Sor</td>\n",
       "      <td>NM_013556</td>\n",
       "      <td>ENSMUSG00000086429</td>\n",
       "      <td>MP:0002873</td>\n",
       "      <td>deltaNLS Wld&lt;S&gt;|deltaNLS&lt;R213A,R215A&gt; Wld&lt;S&gt;</td>\n",
       "      <td>gene trap ROSA 26, Philippe Soriano</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>8989</td>\n",
       "      <td>43851</td>\n",
       "      <td>37740</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1477</td>\n",
       "      <td>1477</td>\n",
       "      <td>247</td>\n",
       "      <td>1477</td>\n",
       "      <td>2068</td>\n",
       "      <td>6</td>\n",
       "      <td>1477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.125652e+07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.454548e+06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.564500e+04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.513958e+07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.085037e+07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.791696e+07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.870304e+07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       mgi_allele_id              allele_symbol  \\\n",
       "count         108940                     108940   \n",
       "unique        108940                     108867   \n",
       "top      MGI:5603349  Arhgap6/Hccs/Mid1<tm1Hzo>   \n",
       "freq               1                          3   \n",
       "mean             NaN                        NaN   \n",
       "std              NaN                        NaN   \n",
       "min              NaN                        NaN   \n",
       "25%              NaN                        NaN   \n",
       "50%              NaN                        NaN   \n",
       "75%              NaN                        NaN   \n",
       "max              NaN                        NaN   \n",
       "\n",
       "                                              allele_name allele_type  \\\n",
       "count                                              108937      108940   \n",
       "unique                                              42229          14   \n",
       "top     endonuclease-mediated mutation 1, GemPharmatec...    Targeted   \n",
       "freq                                                 8989       43851   \n",
       "mean                                                  NaN         NaN   \n",
       "std                                                   NaN         NaN   \n",
       "min                                                   NaN         NaN   \n",
       "25%                                                   NaN         NaN   \n",
       "50%                                                   NaN         NaN   \n",
       "75%                                                   NaN         NaN   \n",
       "max                                                   NaN         NaN   \n",
       "\n",
       "       allele_attribute          PMID mgi_marker_id  marker_symbol  \\\n",
       "count             99813  5.049500e+04        108940         108940   \n",
       "unique             1048           NaN         33118          33115   \n",
       "top       Null/knockout           NaN    MGI:104735  Gt(ROSA)26Sor   \n",
       "freq              37740           NaN          1477           1477   \n",
       "mean                NaN  2.125652e+07           NaN            NaN   \n",
       "std                 NaN  8.454548e+06           NaN            NaN   \n",
       "min                 NaN  2.564500e+04           NaN            NaN   \n",
       "25%                 NaN  1.513958e+07           NaN            NaN   \n",
       "50%                 NaN  2.085037e+07           NaN            NaN   \n",
       "75%                 NaN  2.791696e+07           NaN            NaN   \n",
       "max                 NaN  3.870304e+07           NaN            NaN   \n",
       "\n",
       "       marker_refseq_id   marker_ensembl_id       mp_id  \\\n",
       "count             89676               92229       44935   \n",
       "unique            17298               17741       13572   \n",
       "top           NM_013556  ENSMUSG00000086429  MP:0002873   \n",
       "freq                247                1477        2068   \n",
       "mean                NaN                 NaN         NaN   \n",
       "std                 NaN                 NaN         NaN   \n",
       "min                 NaN                 NaN         NaN   \n",
       "25%                 NaN                 NaN         NaN   \n",
       "50%                 NaN                 NaN         NaN   \n",
       "75%                 NaN                 NaN         NaN   \n",
       "max                 NaN                 NaN         NaN   \n",
       "\n",
       "                                            synonyms  \\\n",
       "count                                          72449   \n",
       "unique                                         70938   \n",
       "top     deltaNLS Wld<S>|deltaNLS<R213A,R215A> Wld<S>   \n",
       "freq                                               6   \n",
       "mean                                             NaN   \n",
       "std                                              NaN   \n",
       "min                                              NaN   \n",
       "25%                                              NaN   \n",
       "50%                                              NaN   \n",
       "75%                                              NaN   \n",
       "max                                              NaN   \n",
       "\n",
       "                                marker_name  \n",
       "count                                108940  \n",
       "unique                                31283  \n",
       "top     gene trap ROSA 26, Philippe Soriano  \n",
       "freq                                   1477  \n",
       "mean                                    NaN  \n",
       "std                                     NaN  \n",
       "min                                     NaN  \n",
       "25%                                     NaN  \n",
       "50%                                     NaN  \n",
       "75%                                     NaN  \n",
       "max                                     NaN  "
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Quick inspection\n",
    "mgi_phenotypic_allele.describe(include='all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "f30fc456-b6c4-4953-95a6-901c3133768f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Targeted', 'Endonuclease-mediated', 'Gene trapped',\n",
       "       'Chemically induced (ENU)', 'Transgenic', 'Spontaneous',\n",
       "       'Not Applicable', 'Transposon induced',\n",
       "       'Chemically induced (other)', 'Radiation induced',\n",
       "       'Chemically and radiation induced', 'QTL', 'Not Specified',\n",
       "       'Other'], dtype=object)"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mgi_phenotypic_allele['allele_type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "a94f8ee4-d2b9-44bd-aea7-2a6028f660bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "108940it [00:13, 7906.16it/s]\n"
     ]
    }
   ],
   "source": [
    "allele_map = {}\n",
    "for idx, row in tqdm(mgi_phenotypic_allele.iterrows()):\n",
    "    d = row.dropna().to_dict()\n",
    "    _id = d.get('mgi_allele_id')\n",
    "    if _id in allele_map:\n",
    "        continue\n",
    "    else:\n",
    "        allele_attribute = d.get('allele_attribute', None)\n",
    "        mammalian_phenotype_id = d.get('mp_id', None)\n",
    "        props = {\n",
    "            'uuid': _id,\n",
    "            'displayName': d.get('allele_symbol', None),\n",
    "            'description': d.get('allele_name', None),\n",
    "            'citations': f'PMID:{d.get(\"PMID\", None)}' if d.get('PMID', None) is not None else None,\n",
    "            'allele_type': d.get('allele_type', None),\n",
    "            'allele_attribute': allele_attribute.split('|') if allele_attribute is not None else None,\n",
    "            'marker_ensembl_id': d.get('marker_ensembl_id'),\n",
    "            'marker_symbol': d.get('marker_symbol'),\n",
    "            'marker_mgi_id': d.get('mgi_marker_id'),\n",
    "            'mammalian_phenotype_id': mammalian_phenotype_id.split(',') if mammalian_phenotype_id is not None else None\n",
    "        }\n",
    "\n",
    "        \n",
    "        n = {\n",
    "            '_id': _id,\n",
    "            'labels': ['Allele'],\n",
    "            'properties': {\n",
    "                k:v for k,v in props.items() if v is not None\n",
    "            }\n",
    "        }\n",
    "        allele_map[_id] = n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11d329d5-0e16-4cdb-a247-d7ab4b5c1b65",
   "metadata": {},
   "source": [
    "## Creating Genotype Nodes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be242c91-ece0-4a82-852f-90dfc55628ea",
   "metadata": {},
   "source": [
    "Next we need to construct all the Genotypes, but also make the edges for the allelic combinations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "73661329-cd2c-4f0a-91fd-7fcdf690fd59",
   "metadata": {},
   "outputs": [],
   "source": [
    "mgi_gene_pheno = download_tab_delimited_data(\n",
    "    'https://www.informatics.jax.org/downloads/reports/MGI_GenePheno.rpt',\n",
    "    ['allelic_composition', 'allele_symbols', 'allele_ids', 'genetic_background', 'mp_id', 'PMID', 'mgi_marker_id', 'mgi_genotype_id']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "b3faeb17-1c3b-4a67-b207-d96678bd762a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allelic_composition</th>\n",
       "      <th>allele_symbols</th>\n",
       "      <th>allele_ids</th>\n",
       "      <th>genetic_background</th>\n",
       "      <th>mp_id</th>\n",
       "      <th>PMID</th>\n",
       "      <th>mgi_marker_id</th>\n",
       "      <th>mgi_genotype_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>MGI:1857242</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0000600</td>\n",
       "      <td>12529408</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>MGI:2166359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>MGI:1857242</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0001716</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>MGI:2166359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>MGI:1857242</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0001698</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>MGI:2166359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>MGI:1857242</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0001092</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>MGI:2166359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>MGI:1857242</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0000961</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>MGI:2166359</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       allelic_composition allele_symbols   allele_ids     genetic_background  \\\n",
       "0  Rb1<tm1Tyj>/Rb1<tm1Tyj>    Rb1<tm1Tyj>  MGI:1857242  involves: 129S2/SvPas   \n",
       "1  Rb1<tm1Tyj>/Rb1<tm1Tyj>    Rb1<tm1Tyj>  MGI:1857242  involves: 129S2/SvPas   \n",
       "2  Rb1<tm1Tyj>/Rb1<tm1Tyj>    Rb1<tm1Tyj>  MGI:1857242  involves: 129S2/SvPas   \n",
       "3  Rb1<tm1Tyj>/Rb1<tm1Tyj>    Rb1<tm1Tyj>  MGI:1857242  involves: 129S2/SvPas   \n",
       "4  Rb1<tm1Tyj>/Rb1<tm1Tyj>    Rb1<tm1Tyj>  MGI:1857242  involves: 129S2/SvPas   \n",
       "\n",
       "        mp_id      PMID mgi_marker_id mgi_genotype_id  \n",
       "0  MP:0000600  12529408     MGI:97874     MGI:2166359  \n",
       "1  MP:0001716  16449662     MGI:97874     MGI:2166359  \n",
       "2  MP:0001698  16449662     MGI:97874     MGI:2166359  \n",
       "3  MP:0001092  16449662     MGI:97874     MGI:2166359  \n",
       "4  MP:0000961  16449662     MGI:97874     MGI:2166359  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mgi_gene_pheno.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "21193816-2002-49ac-b9ca-29ce17604086",
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype_map = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "27f013ed-70f4-47e3-8c51-6db69a3e8a5a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "48999it [00:02, 21341.51it/s]\n"
     ]
    }
   ],
   "source": [
    "genotype_subset = mgi_gene_pheno[['mgi_genotype_id', 'allelic_composition', 'genetic_background']].groupby('mgi_genotype_id').first()\n",
    "for _id, row in tqdm(genotype_subset.iterrows()):\n",
    "    d = row.to_dict()\n",
    "    if _id in genotype_map:\n",
    "        continue\n",
    "    else:\n",
    "        n = {\n",
    "            '_id': _id,\n",
    "            'labels': ['Genotype'],\n",
    "            'properties': {\n",
    "                'uuid': _id,\n",
    "                'displayName': d.get('allelic_composition'),\n",
    "                'description': '',\n",
    "                'allelic_composition': d.get('allelic_composition'),\n",
    "                'genetic_background': d.get('genetic_background'),\n",
    "                'url': f\"https://www.informatics.jax.org/allele/genoview/{_id}\"\n",
    "            }\n",
    "        }\n",
    "        genotype_map[_id] = n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "b7942e95-ea5d-411c-bb12-969f5e40f541",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset out allele_id and genotype_id\n",
    "geno2allele = mgi_gene_pheno[['allele_ids', 'mgi_genotype_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "d0514c7d-7b6f-431a-b7e0-1b42756e4929",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allele_ids</th>\n",
       "      <th>mgi_genotype_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGI:1857242</td>\n",
       "      <td>MGI:2166359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MGI:1857242</td>\n",
       "      <td>MGI:2166359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:1857242</td>\n",
       "      <td>MGI:2166359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MGI:1857242</td>\n",
       "      <td>MGI:2166359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MGI:1857242</td>\n",
       "      <td>MGI:2166359</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    allele_ids mgi_genotype_id\n",
       "0  MGI:1857242     MGI:2166359\n",
       "1  MGI:1857242     MGI:2166359\n",
       "2  MGI:1857242     MGI:2166359\n",
       "3  MGI:1857242     MGI:2166359\n",
       "4  MGI:1857242     MGI:2166359"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geno2allele.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "caa81350-3227-49b6-858c-e166a80f2832",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "geno2allele_edges = []\n",
    "\n",
    "for name, group in geno2allele.groupby('mgi_genotype_id'):\n",
    "    allele_ids = set()\n",
    "    for allele_str in list(group['allele_ids'].unique()):\n",
    "        allele_ids.update(allele_str.split('|'))\n",
    "    \n",
    "    \n",
    "    for allele_id in list(allele_ids):\n",
    "        e = {\n",
    "            'from': {\n",
    "                'uuid': name\n",
    "            },\n",
    "            'to': {\n",
    "                'uuid': allele_id\n",
    "            },\n",
    "            'label': 'has allele',\n",
    "            'properties': {}\n",
    "        }\n",
    "        geno2allele_edges.append(e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "1bd7424b-dee0-44a8-a24a-dd280c02365f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'from': {'uuid': 'MGI:2166359'},\n",
       " 'to': {'uuid': 'MGI:1857242'},\n",
       " 'label': 'has allele',\n",
       " 'properties': {}}"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geno2allele_edges[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fd4863d-e9bc-45cc-9dd3-1badd8cec588",
   "metadata": {},
   "source": [
    "Convert the DOIDs to EFOs where possible, and then explode out the genotypes for edges. Assumes EFO and MONDO are in."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "3c6893af-47c2-4a67-8559-dd73c272db30",
   "metadata": {},
   "outputs": [],
   "source": [
    "mgi_disease_gene_model = download_tab_delimited_data('https://www.informatics.jax.org/downloads/reports/MGI_DiseaseGeneModel.rpt',\n",
    "                                                    [\n",
    "                                                        'human_gene_symbol',\n",
    "                                                        'human_gene_name',\n",
    "                                                        'hgnc_id',\n",
    "                                                        'do_term',\n",
    "                                                        'do_term_id',\n",
    "                                                        'mouse_genotype_id',\n",
    "                                                        'mouse_gene',\n",
    "                                                        'mouse_mgi_id',\n",
    "                                                        'facilities',\n",
    "                                                        'repo_id'\n",
    "                                                    ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "4697593b-274c-4566-b17f-9dffb5367bef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>human_gene_symbol</th>\n",
       "      <th>human_gene_name</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>do_term</th>\n",
       "      <th>do_term_id</th>\n",
       "      <th>mouse_genotype_id</th>\n",
       "      <th>mouse_gene</th>\n",
       "      <th>mouse_mgi_id</th>\n",
       "      <th>facilities</th>\n",
       "      <th>repo_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>HGNC:24086</td>\n",
       "      <td>gout</td>\n",
       "      <td>DOID:13189</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A1cf</td>\n",
       "      <td>MGI:1917115</td>\n",
       "      <td>APB|EMMA|EuMMCR|GPT|JAX|MMRRC</td>\n",
       "      <td>EuMMCR:8829|EuMMCR:8828|EM:11705|APB:7708|MMRR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>Alzheimer's disease</td>\n",
       "      <td>DOID:10652</td>\n",
       "      <td>MGI:3693439|MGI:2684657|MGI:3721938|MGI:575438...</td>\n",
       "      <td>A2m</td>\n",
       "      <td>MGI:2449119</td>\n",
       "      <td>CMMR|GPT|MMRRC|SMOC</td>\n",
       "      <td>CMMR:CMMR:ABAL|CMMR:CMMR:ABDJ|CMMR:CMMR:ABHA|N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>background diabetic retinopathy</td>\n",
       "      <td>DOID:13208</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A2m</td>\n",
       "      <td>MGI:2449119</td>\n",
       "      <td>CMMR|GPT|MMRRC|SMOC</td>\n",
       "      <td>CMMR:CMMR:ABAL|CMMR:CMMR:ABDJ|CMMR:CMMR:ABHA|N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>DOID:0080600</td>\n",
       "      <td>MGI:6402641|MGI:6455041|MGI:6431340|MGI:6415340</td>\n",
       "      <td>A2m</td>\n",
       "      <td>MGI:2449119</td>\n",
       "      <td>CMMR|GPT|MMRRC|SMOC</td>\n",
       "      <td>CMMR:CMMR:ABAL|CMMR:CMMR:ABDJ|CMMR:CMMR:ABHA|N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>multiple sclerosis</td>\n",
       "      <td>DOID:2377</td>\n",
       "      <td>MGI:3835028|MGI:6305814|MGI:4461062|MGI:542970...</td>\n",
       "      <td>A2m</td>\n",
       "      <td>MGI:2449119</td>\n",
       "      <td>CMMR|GPT|MMRRC|SMOC</td>\n",
       "      <td>CMMR:CMMR:ABAL|CMMR:CMMR:ABDJ|CMMR:CMMR:ABHA|N...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  human_gene_symbol                 human_gene_name     hgnc_id  \\\n",
       "0              A1CF  APOBEC1 complementation factor  HGNC:24086   \n",
       "1               A2M           alpha-2-macroglobulin      HGNC:7   \n",
       "2               A2M           alpha-2-macroglobulin      HGNC:7   \n",
       "3               A2M           alpha-2-macroglobulin      HGNC:7   \n",
       "4               A2M           alpha-2-macroglobulin      HGNC:7   \n",
       "\n",
       "                           do_term    do_term_id  \\\n",
       "0                             gout    DOID:13189   \n",
       "1              Alzheimer's disease    DOID:10652   \n",
       "2  background diabetic retinopathy    DOID:13208   \n",
       "3                         COVID-19  DOID:0080600   \n",
       "4               multiple sclerosis     DOID:2377   \n",
       "\n",
       "                                   mouse_genotype_id mouse_gene mouse_mgi_id  \\\n",
       "0                                                NaN       A1cf  MGI:1917115   \n",
       "1  MGI:3693439|MGI:2684657|MGI:3721938|MGI:575438...        A2m  MGI:2449119   \n",
       "2                                                NaN        A2m  MGI:2449119   \n",
       "3    MGI:6402641|MGI:6455041|MGI:6431340|MGI:6415340        A2m  MGI:2449119   \n",
       "4  MGI:3835028|MGI:6305814|MGI:4461062|MGI:542970...        A2m  MGI:2449119   \n",
       "\n",
       "                      facilities  \\\n",
       "0  APB|EMMA|EuMMCR|GPT|JAX|MMRRC   \n",
       "1            CMMR|GPT|MMRRC|SMOC   \n",
       "2            CMMR|GPT|MMRRC|SMOC   \n",
       "3            CMMR|GPT|MMRRC|SMOC   \n",
       "4            CMMR|GPT|MMRRC|SMOC   \n",
       "\n",
       "                                             repo_id  \n",
       "0  EuMMCR:8829|EuMMCR:8828|EM:11705|APB:7708|MMRR...  \n",
       "1  CMMR:CMMR:ABAL|CMMR:CMMR:ABDJ|CMMR:CMMR:ABHA|N...  \n",
       "2  CMMR:CMMR:ABAL|CMMR:CMMR:ABDJ|CMMR:CMMR:ABHA|N...  \n",
       "3  CMMR:CMMR:ABAL|CMMR:CMMR:ABDJ|CMMR:CMMR:ABHA|N...  \n",
       "4  CMMR:CMMR:ABAL|CMMR:CMMR:ABDJ|CMMR:CMMR:ABHA|N...  "
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mgi_disease_gene_model.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "b81a3c79-51a2-4f8d-9dcc-53a2eebfda32",
   "metadata": {},
   "outputs": [],
   "source": [
    "mgi_disease_gene_model.dropna(subset=['mouse_genotype_id'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "1ca40ce3-b468-4084-8538-b7d83c33d135",
   "metadata": {},
   "outputs": [],
   "source": [
    "doid_grouped = mgi_disease_gene_model[['do_term_id', 'mouse_genotype_id']].groupby('do_term_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "8ed78236-f2c4-4504-971f-562eb5b009c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./resources/mappings/doid_efo.mapping.json', 'r') as f:\n",
    "    doid_efo_mapping = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "bae75cd8-64e3-4e39-86b4-2e64d89f411a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6127"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(doid_efo_mapping.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "cd36620e-469a-4f26-9708-85d1931e3b6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype2disease_edges = []\n",
    "\n",
    "for doid, group in doid_grouped:\n",
    "    if doid not in doid_efo_mapping:\n",
    "        continue\n",
    "    else:\n",
    "        \n",
    "        efo_hits = doid_efo_mapping[doid]\n",
    "        if efo_hits is None:\n",
    "            continue\n",
    "        efo_id = efo_hits[0]\n",
    "        genotype_ids = set()\n",
    "        for genotype_str in list(group['mouse_genotype_id']):\n",
    "            genotype_ids.update(genotype_str.split('|'))\n",
    "\n",
    "        for genotype_id in list(genotype_ids):\n",
    "            e = {\n",
    "                'from': {\n",
    "                    'uuid': genotype_id\n",
    "                },\n",
    "                'to': {\n",
    "                    'uuid': efo_id\n",
    "                },\n",
    "                'label': 'is model of',\n",
    "                'properties': {}\n",
    "            }\n",
    "            genotype2disease_edges.append(e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "fb4b73ff-c849-4767-a16d-67dedf0a23e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'from': {'uuid': 'MGI:5314017'},\n",
       " 'to': {'uuid': 'EFO:1001352'},\n",
       " 'label': 'is model of',\n",
       " 'properties': {}}"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype2disease_edges[10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de716666-32bd-4e2a-9609-906161ab975b",
   "metadata": {},
   "source": [
    "# Node and Edge Generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "72b5a773-0ba1-492d-b89a-12da15bbc69d",
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes = []\n",
    "edges = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "4f097893-89f4-40c2-b3a6-3debe83511be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48999\n",
      "50361\n"
     ]
    }
   ],
   "source": [
    "nodes.extend(genotype_map.values())\n",
    "print(len(nodes))\n",
    "edges.extend(geno2allele_edges)\n",
    "print(len(edges))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "0c3b56a2-8dce-4cd8-99b1-c1dc9bbd1f8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "157939\n",
      "57477\n"
     ]
    }
   ],
   "source": [
    "nodes.extend(allele_map.values())\n",
    "print(len(nodes))\n",
    "edges.extend(genotype2disease_edges)\n",
    "print(len(edges))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "3ec342ab-9694-4150-8bed-ea615e2a5a21",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 157939/157939 [00:03<00:00, 45774.97it/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 57477/57477 [00:00<00:00, 122554.33it/s]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "with gzip.open(os.path.join(processed_data, 'alliance_disease_models_node.jsonl.gz'), 'wt', encoding='utf-8') as outfile:\n",
    "    for x in tqdm(nodes):\n",
    "        outfile.write(json.dumps(x) + '\\n')\n",
    "\n",
    "with gzip.open(os.path.join(processed_data, 'alliance_disease_models_edge.jsonl.gz'), 'wt', encoding='utf-8') as outfile:\n",
    "    for x in tqdm(edges):\n",
    "        outfile.write(json.dumps(x) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "80f94a6c-269c-4bf0-a5ca-da67097bfe7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = {\n",
    "    \"_meta\": {\n",
    "        \"version\": \"0.1.3\",\n",
    "        \"date_updated\": \"2024-02-17\",\n",
    "        \"maintainer\": \"BioBox Analytics\"\n",
    "    },\n",
    "    \"key\": \"alliance_disease_models\",\n",
    "    \"name\": \"Alliance Genome - Animal Models of Disease\",\n",
    "    \"description\": \"Extraction of alleles-genotypes-diseases reports from MGI.\",\n",
    "    \"source\": [\n",
    "        {\n",
    "            \"uri\": \"https://www.informatics.jax.org/\",\n",
    "            \"type\": \"doc\"\n",
    "        },\n",
    "        {\n",
    "            \"uri\": \"https://www.informatics.jax.org/downloads/reports/index.html#pheno\",\n",
    "            \"type\": \"data\",\n",
    "            \"version\": \"7.1.0\"\n",
    "        }\n",
    "    ],\n",
    "    \"concepts\": {\n",
    "        \"Allele\": {\n",
    "            \"label\": \"Allele\",\n",
    "            \"dbLabel\": \"Allele\",\n",
    "            \"definition\": \"One of a set of sequence features known to exist at a particular genomic location. An allele is a seqeunce feature at a genomic location where variation occurs (i.e. where >1 different sequence is known to exist). An allele can span only the extent of sequence known to vary (e.g. a single base SNP, or short insertion), or it can span a larger extent that includes one or more variable features as proper parts (e.g. a 'gene allele' that spans the extent of an entire gene which contains several sequence alterations). Alleles can carry 'reference' or 'variant' sequence - depending on whether the its 'state' matches that considered to be the reference at that location. Alleles whose state differs from the reference are called 'variant alleles', and those that match the reference are called 'reference alleles'. What is considered the 'reference' state at a particular location may vary, depending on the context/goal of a particular analysis. A 'sequence alteration' is a 'variant allele' that varies along its entire extent (i.e every position varies from that of some defined reference sequence).\"\n",
    "        },\n",
    "        \"Genotype\": {\n",
    "            \"label\": \"Genotype\",\n",
    "            \"dbLabel\": \"Genotype\",\n",
    "            \"definition\": \"The total sum of the genetic information of an organism that is known and relevant to the experiment being performed, including chromosomal, plasmid, viral or other genetic material which has been introduced into the organism either prior to or during the experiment. Information, making the distinction between the actual physical material (e.g. a cell) and the information about the genetic content (genotype).\"\n",
    "        }\n",
    "    },\n",
    "    \"relationships\": {\n",
    "        \"has allele\": {\n",
    "            \"from\": \"Genotype\",\n",
    "            \"to\": \"Allele\"\n",
    "        },\n",
    "        \"is model of\": {\n",
    "            \"from\": \"Genotype\",\n",
    "            \"to\": \"Disease\"\n",
    "        }\n",
    "    }\n",
    "}\n",
    "with open(os.path.join(processed_data, 'alliance_disease_models_metadata.json'), 'w') as f:\n",
    "    json.dump(metadata, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26fab6af-24e0-4704-97cc-1560b4f404cf",
   "metadata": {},
   "source": [
    "# Scratch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "58165ed6-4bcf-4e9c-899b-ab1e27ed79b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from neo4j import GraphDatabase\n",
    "\n",
    "driver = GraphDatabase.driver(\"bolt+ssc://pr-1645-neo4j-pr-1645.frzb.biobox.to:7687\", auth=('neo4j', 'EpPRhKPyJUxgC6gv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "04e1867d-b236-465e-8239-e5c147da78e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = requests.get('http://www.ensembl.org/biomart/martservice?query=<?xml version=\"1.0\" encoding=\"UTF-8\"?><!DOCTYPE Query><Query  virtualSchemaName = \"default\" formatter = \"TSV\" header = \"0\" uniqueRows = \"0\" count = \"\" datasetConfigVersion = \"0.6\" ><Dataset name = \"hsapiens_gene_ensembl\" interface = \"default\" ><Attribute name = \"ensembl_gene_id\" /><Attribute name = \"mmusculus_homolog_ensembl_gene\" /></Dataset></Query>')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "853cdaba-6ba4-4980-ba3c-6896c775bcaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_mouse_homolog = pd.read_table(io.StringIO(res.text), header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4464a05c-f461-490d-9591-86ca7d0266b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_mouse_homolog.columns = ['hsapiens', 'mmusculus']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "18db958d-ff9a-4de2-9317-ecf723b011c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_mouse_homolog.dropna(subset='mmusculus', inplace=True)\n",
    "human_mouse_homolog.set_index('mmusculus', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b27878a6-0938-4f91-8791-042d4b0842dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ENSG00000198888'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_mouse_homolog.loc['ENSMUSG00000064341']['hsapiens']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2afd25c9-59f8-43f2-9737-233160d42aad",
   "metadata": {},
   "outputs": [],
   "source": [
    "records, summary, _ = driver.execute_query(\n",
    "    \"\"\"\n",
    "    MATCH (a:Allele)\n",
    "    WHERE a.marker_ensembl_id is not null\n",
    "    RETURN a.uuid, a.marker_ensembl_id\n",
    "    \"\"\",\n",
    "    database_=\"bx-d52b76a6-2e9b-4d53-894e-58e818cff0f2\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f4ab1619-a0a4-475f-b8b4-3e0f118231b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "batch = []\n",
    "\n",
    "for r in records:\n",
    "    allele_id = r.get('a.uuid')\n",
    "    mouse_gene_id = r.get('a.marker_ensembl_id')\n",
    " \n",
    "    try:\n",
    "        hsapien_id = human_mouse_homolog.loc[mouse_gene_id]['hsapiens']\n",
    "        batch.append({'source': allele_id, 'target': hsapien_id})\n",
    "    except:\n",
    "        continue\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a76eb582-54cb-4657-883a-47df0746d63a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "87157"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(batch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "381bafd0-b496-420b-b89a-51fc4e0b2fc2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "EagerResult(records=[], summary=<neo4j._work.summary.ResultSummary object at 0x12f856490>, keys=[])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "driver.execute_query(\n",
    "    \"\"\"\n",
    "    UNWIND $batch as b\n",
    "    MATCH (a:Allele {uuid: b.source})\n",
    "    MATCH (g:Gene { uuid: b.target})\n",
    "    MERGE (a)-[:`on target`]->(g)\n",
    "    \"\"\",\n",
    "    database_=\"bx-d52b76a6-2e9b-4d53-894e-58e818cff0f2\",\n",
    "    batch=batch\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "bcc39206-270c-4f77-bd17-b2acff2f2e4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = requests.post(\"https://pr-1625-rest.frzb.biobox.to/bioref/query/objects\",\n",
    "                   json={ \"query\": \"(Genotype)-> is model of ->(Disease)<- is implicated in severity of <-(Gene)\", \"class\": \"Gene\", \"pagination\": { \"limit\": 100, \"offset\": 0, \"sortBy\": { \"key\": \"uuid\", \"descending\": False } } },\n",
    "                   headers= {\n",
    "                       'x-biobox-orgid': 'bx-3079fc5a-1ed1-4e17-a482-e38efdbfe837',\n",
    "                        'Authorization': f\"Bearer {}\"\n",
    "                   })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "514b4d06-4fb2-4fb3-bbc0-2d46938d1100",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'class': 'Gene',\n",
       "  'variable': 'g2',\n",
       "  'statement': 'MATCH\\n(g:`Genotype`)-[_r:`is model of`]->(d:`Disease`)<-[_r2:`is implicated in severity of`]-(g2:`Gene`)RETURN DISTINCT g2\\nSKIP 0\\nLIMIT 100',\n",
       "  'countStatement': 'MATCH\\n(g:`Genotype`)-[_r:`is model of`]->(d:`Disease`)<-[_r2:`is implicated in severity of`]-(g2:`Gene`)RETURN count(DISTINCT g2)',\n",
       "  'total': 327,\n",
       "  'offset': 0,\n",
       "  'count': 100,\n",
       "  'data': [{'biotype': 'protein_coding',\n",
       "    'symbol': 'ALX4',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'ALX4',\n",
       "    'start': 44260440,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '11',\n",
       "    'uuid': 'ENSG00000052850',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000052850.8',\n",
       "    'HGNC': 'HGNC:450',\n",
       "    'end': 44310139,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'ARID1A',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'ARID1A',\n",
       "    'start': 26693236,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '1',\n",
       "    'uuid': 'ENSG00000117713',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000117713.21',\n",
       "    'HGNC': 'HGNC:11110',\n",
       "    'end': 26782104,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'ARID2',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'ARID2',\n",
       "    'start': 45729706,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '12',\n",
       "    'uuid': 'ENSG00000189079',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000189079.18',\n",
       "    'HGNC': 'HGNC:18037',\n",
       "    'end': 45908040,\n",
       "    'tag': ['ncRNA_host', 'overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'BIRC3',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'BIRC3',\n",
       "    'start': 102317484,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '11',\n",
       "    'uuid': 'ENSG00000023445',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000023445.17',\n",
       "    'HGNC': 'HGNC:591',\n",
       "    'end': 102339403,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'CASP8',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'CASP8',\n",
       "    'start': 201233443,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '2',\n",
       "    'uuid': 'ENSG00000064012',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000064012.24',\n",
       "    'HGNC': 'HGNC:1509',\n",
       "    'end': 201361836,\n",
       "    'tag': ['overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'CSF1R',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'CSF1R',\n",
       "    'start': 150053291,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '5',\n",
       "    'uuid': 'ENSG00000182578',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000182578.14',\n",
       "    'HGNC': 'HGNC:2433',\n",
       "    'end': 150113372,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'CTNNB1',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'CTNNB1',\n",
       "    'start': 41194741,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '3',\n",
       "    'uuid': 'ENSG00000168036',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000168036.18',\n",
       "    'HGNC': 'HGNC:2514',\n",
       "    'end': 41260096,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'EFTUD2',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'EFTUD2',\n",
       "    'start': 44849948,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '17',\n",
       "    'uuid': 'ENSG00000108883',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000108883.13',\n",
       "    'HGNC': 'HGNC:30858',\n",
       "    'end': 44899445,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'FES',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'FES',\n",
       "    'start': 90883695,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '15',\n",
       "    'uuid': 'ENSG00000182511',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000182511.12',\n",
       "    'HGNC': 'HGNC:3657',\n",
       "    'end': 90895776,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'FGFR2',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'FGFR2',\n",
       "    'start': 121478332,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '10',\n",
       "    'uuid': 'ENSG00000066468',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000066468.24',\n",
       "    'HGNC': 'HGNC:3689',\n",
       "    'end': 121598458,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'FLT3',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'FLT3',\n",
       "    'start': 28003274,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '13',\n",
       "    'uuid': 'ENSG00000122025',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000122025.15',\n",
       "    'HGNC': 'HGNC:3765',\n",
       "    'end': 28100592,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'FUBP1',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'FUBP1',\n",
       "    'start': 77944055,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '1',\n",
       "    'uuid': 'ENSG00000162613',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000162613.18',\n",
       "    'HGNC': 'HGNC:4004',\n",
       "    'end': 77979110,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'GPX1',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'GPX1',\n",
       "    'start': 49357174,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '3',\n",
       "    'uuid': 'ENSG00000233276',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000233276.9',\n",
       "    'HGNC': 'HGNC:4553',\n",
       "    'end': 49358605,\n",
       "    'tag': ['overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'HNRNPAB',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'HNRNPAB',\n",
       "    'start': 178204533,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '5',\n",
       "    'uuid': 'ENSG00000197451',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000197451.12',\n",
       "    'HGNC': 'HGNC:5034',\n",
       "    'end': 178211163,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'IL18',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'IL18',\n",
       "    'start': 112143253,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '11',\n",
       "    'uuid': 'ENSG00000150782',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000150782.12',\n",
       "    'HGNC': 'HGNC:5986',\n",
       "    'end': 112164096,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'KDM6A',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'KDM6A',\n",
       "    'start': 44873188,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': 'X',\n",
       "    'uuid': 'ENSG00000147050',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000147050.18',\n",
       "    'HGNC': 'HGNC:12637',\n",
       "    'end': 45112779,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'LECT2',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'LECT2',\n",
       "    'start': 135922279,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '5',\n",
       "    'uuid': 'ENSG00000145826',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000145826.9',\n",
       "    'HGNC': 'HGNC:6550',\n",
       "    'end': 135954983,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'LRP1B',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'LRP1B',\n",
       "    'start': 140231423,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '2',\n",
       "    'uuid': 'ENSG00000168702',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000168702.18',\n",
       "    'HGNC': 'HGNC:6693',\n",
       "    'end': 142131016,\n",
       "    'tag': ['ncRNA_host', 'overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'MAP4K4',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'MAP4K4',\n",
       "    'start': 101696850,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '2',\n",
       "    'uuid': 'ENSG00000071054',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000071054.17',\n",
       "    'HGNC': 'HGNC:6866',\n",
       "    'end': 101894690,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'MBL2',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'MBL2',\n",
       "    'start': 52765380,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '10',\n",
       "    'uuid': 'ENSG00000165471',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000165471.7',\n",
       "    'HGNC': 'HGNC:6922',\n",
       "    'end': 52772784,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'miRNA',\n",
       "    'symbol': 'MIR195',\n",
       "    'level': 'automatically annotated loci',\n",
       "    'displayName': 'MIR195',\n",
       "    'start': 7017615,\n",
       "    'source': 'ENSEMBL',\n",
       "    'chr': '17',\n",
       "    'uuid': 'ENSG00000284112',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000284112.1',\n",
       "    'HGNC': 'HGNC:31566',\n",
       "    'end': 7017701,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'miRNA',\n",
       "    'symbol': 'MIR212',\n",
       "    'level': 'automatically annotated loci',\n",
       "    'displayName': 'MIR212',\n",
       "    'start': 2050271,\n",
       "    'source': 'ENSEMBL',\n",
       "    'chr': '17',\n",
       "    'uuid': 'ENSG00000267195',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000267195.1',\n",
       "    'HGNC': 'HGNC:31589',\n",
       "    'end': 2050380,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'miRNA',\n",
       "    'symbol': 'MIR27A',\n",
       "    'level': 'automatically annotated loci',\n",
       "    'displayName': 'MIR27A',\n",
       "    'start': 13836440,\n",
       "    'source': 'ENSEMBL',\n",
       "    'chr': '19',\n",
       "    'uuid': 'ENSG00000207808',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000207808.1',\n",
       "    'HGNC': 'HGNC:31613',\n",
       "    'end': 13836517,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'miRNA',\n",
       "    'symbol': 'MIR302D',\n",
       "    'level': 'automatically annotated loci',\n",
       "    'displayName': 'MIR302D',\n",
       "    'start': 112648004,\n",
       "    'source': 'ENSEMBL',\n",
       "    'chr': '4',\n",
       "    'uuid': 'ENSG00000199145',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000199145.1',\n",
       "    'HGNC': 'HGNC:31765',\n",
       "    'end': 112648071,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'MYD88',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'MYD88',\n",
       "    'start': 38138552,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '3',\n",
       "    'uuid': 'ENSG00000172936',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000172936.18',\n",
       "    'HGNC': 'HGNC:7562',\n",
       "    'end': 38143024,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'NRAS',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'NRAS',\n",
       "    'start': 114704469,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '1',\n",
       "    'uuid': 'ENSG00000213281',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000213281.5',\n",
       "    'HGNC': 'HGNC:7989',\n",
       "    'end': 114716771,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'PIK3CA',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'PIK3CA',\n",
       "    'start': 179148114,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '3',\n",
       "    'uuid': 'ENSG00000121879',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000121879.6',\n",
       "    'HGNC': 'HGNC:8975',\n",
       "    'end': 179240093,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'PLCG1',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'PLCG1',\n",
       "    'start': 41136960,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '20',\n",
       "    'uuid': 'ENSG00000124181',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000124181.15',\n",
       "    'HGNC': 'HGNC:9065',\n",
       "    'end': 41196801,\n",
       "    'tag': ['ncRNA_host', 'overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'RIT1',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'RIT1',\n",
       "    'start': 155897808,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '1',\n",
       "    'uuid': 'ENSG00000143622',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000143622.12',\n",
       "    'HGNC': 'HGNC:10023',\n",
       "    'end': 155911404,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'ENSG00000285897',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'ENSG00000285897',\n",
       "    'start': 58345206,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '17',\n",
       "    'uuid': 'ENSG00000285897',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000285897.1',\n",
       "    'HGNC': '',\n",
       "    'end': 58417595,\n",
       "    'tag': ['overlapping_locus', 'readthrough_gene']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'SLC38A1',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'SLC38A1',\n",
       "    'start': 46183063,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '12',\n",
       "    'uuid': 'ENSG00000111371',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000111371.16',\n",
       "    'HGNC': 'HGNC:13447',\n",
       "    'end': 46270017,\n",
       "    'tag': ['overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'SLC7A5',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'SLC7A5',\n",
       "    'start': 87830023,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '16',\n",
       "    'uuid': 'ENSG00000103257',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000103257.9',\n",
       "    'HGNC': 'HGNC:11063',\n",
       "    'end': 87869507,\n",
       "    'tag': ['ncRNA_host']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'SMC3',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'SMC3',\n",
       "    'start': 110567684,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '10',\n",
       "    'uuid': 'ENSG00000108055',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000108055.11',\n",
       "    'HGNC': 'HGNC:2468',\n",
       "    'end': 110606048,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'SOX30',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'SOX30',\n",
       "    'start': 157625679,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '5',\n",
       "    'uuid': 'ENSG00000039600',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000039600.11',\n",
       "    'HGNC': 'HGNC:30635',\n",
       "    'end': 157671480,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'SPOCK1',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'SPOCK1',\n",
       "    'start': 136975298,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '5',\n",
       "    'uuid': 'ENSG00000152377',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000152377.15',\n",
       "    'HGNC': 'HGNC:11251',\n",
       "    'end': 137598379,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'SPOP',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'SPOP',\n",
       "    'start': 49598884,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '17',\n",
       "    'uuid': 'ENSG00000121067',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000121067.19',\n",
       "    'HGNC': 'HGNC:11254',\n",
       "    'end': 49678163,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'TET2',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'TET2',\n",
       "    'start': 105145875,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '4',\n",
       "    'uuid': 'ENSG00000168769',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000168769.14',\n",
       "    'HGNC': 'HGNC:25941',\n",
       "    'end': 105279816,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'XPO1',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'XPO1',\n",
       "    'start': 61476032,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '2',\n",
       "    'uuid': 'ENSG00000082898',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000082898.19',\n",
       "    'HGNC': 'HGNC:12825',\n",
       "    'end': 61538741,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'XPO5',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'XPO5',\n",
       "    'start': 43522334,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '6',\n",
       "    'uuid': 'ENSG00000124571',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000124571.18',\n",
       "    'HGNC': 'HGNC:17675',\n",
       "    'end': 43576038,\n",
       "    'tag': ['overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'XRCC1',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'XRCC1',\n",
       "    'start': 43543311,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '19',\n",
       "    'uuid': 'ENSG00000073050',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000073050.12',\n",
       "    'HGNC': 'HGNC:12828',\n",
       "    'end': 43580473,\n",
       "    'tag': ['overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'ZFHX3',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'ZFHX3',\n",
       "    'start': 72782885,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '16',\n",
       "    'uuid': 'ENSG00000140836',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000140836.17',\n",
       "    'HGNC': 'HGNC:777',\n",
       "    'end': 73891871,\n",
       "    'tag': ['ncRNA_host']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'ADAM28',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'ADAM28',\n",
       "    'start': 24294069,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '8',\n",
       "    'uuid': 'ENSG00000042980',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000042980.14',\n",
       "    'HGNC': 'HGNC:206',\n",
       "    'end': 24359014,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'CADM1',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'CADM1',\n",
       "    'start': 115169218,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '11',\n",
       "    'uuid': 'ENSG00000182985',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000182985.19',\n",
       "    'HGNC': 'HGNC:5951',\n",
       "    'end': 115504957,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'CBX7',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'CBX7',\n",
       "    'start': 39120167,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '22',\n",
       "    'uuid': 'ENSG00000100307',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000100307.13',\n",
       "    'HGNC': 'HGNC:1557',\n",
       "    'end': 39152680,\n",
       "    'tag': ['overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'DUSP1',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'DUSP1',\n",
       "    'start': 172768096,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '5',\n",
       "    'uuid': 'ENSG00000120129',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000120129.6',\n",
       "    'HGNC': 'HGNC:3064',\n",
       "    'end': 172771195,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'FBXO11',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'FBXO11',\n",
       "    'start': 47789316,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '2',\n",
       "    'uuid': 'ENSG00000138081',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000138081.22',\n",
       "    'HGNC': 'HGNC:13590',\n",
       "    'end': 47906498,\n",
       "    'tag': ['overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'GNRH1',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'GNRH1',\n",
       "    'start': 25419258,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '8',\n",
       "    'uuid': 'ENSG00000147437',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000147437.10',\n",
       "    'HGNC': 'HGNC:4419',\n",
       "    'end': 25424654,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'ENSG00000293552',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'ENSG00000293552',\n",
       "    'start': 6716157,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '9',\n",
       "    'uuid': 'ENSG00000293552',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000293552.1',\n",
       "    'HGNC': '',\n",
       "    'end': 7077263,\n",
       "    'tag': ['overlapping_locus', 'readthrough_gene']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'KSR1',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'KSR1',\n",
       "    'start': 27456448,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '17',\n",
       "    'uuid': 'ENSG00000141068',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000141068.17',\n",
       "    'HGNC': 'HGNC:6465',\n",
       "    'end': 27626438,\n",
       "    'tag': ['overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'LHCGR',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'LHCGR',\n",
       "    'start': 48686774,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '2',\n",
       "    'uuid': 'ENSG00000138039',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000138039.15',\n",
       "    'HGNC': 'HGNC:6585',\n",
       "    'end': 48755730,\n",
       "    'tag': ['overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'MCPH1',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'MCPH1',\n",
       "    'start': 6406592,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '8',\n",
       "    'uuid': 'ENSG00000147316',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000147316.15',\n",
       "    'HGNC': 'HGNC:6954',\n",
       "    'end': 6648508,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'MYC',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'MYC',\n",
       "    'start': 127735434,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '8',\n",
       "    'uuid': 'ENSG00000136997',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000136997.22',\n",
       "    'HGNC': 'HGNC:7553',\n",
       "    'end': 127742951,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'TGFB1',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'TGFB1',\n",
       "    'start': 41301587,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '19',\n",
       "    'uuid': 'ENSG00000105329',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000105329.11',\n",
       "    'HGNC': 'HGNC:11766',\n",
       "    'end': 41353922,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'TNFAIP3',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'TNFAIP3',\n",
       "    'start': 137867214,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '6',\n",
       "    'uuid': 'ENSG00000118503',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000118503.18',\n",
       "    'HGNC': 'HGNC:11896',\n",
       "    'end': 137883314,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'TSG101',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'TSG101',\n",
       "    'start': 18468336,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '11',\n",
       "    'uuid': 'ENSG00000074319',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000074319.13',\n",
       "    'HGNC': 'HGNC:15971',\n",
       "    'end': 18526951,\n",
       "    'tag': ['overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'TWIST1',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'TWIST1',\n",
       "    'start': 19020991,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '7',\n",
       "    'uuid': 'ENSG00000122691',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000122691.13',\n",
       "    'HGNC': 'HGNC:12428',\n",
       "    'end': 19117636,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'UBR5',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'UBR5',\n",
       "    'start': 102252273,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '8',\n",
       "    'uuid': 'ENSG00000104517',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000104517.13',\n",
       "    'HGNC': 'HGNC:16806',\n",
       "    'end': 102412759,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'CDKN2A',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'CDKN2A',\n",
       "    'start': 21967752,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '9',\n",
       "    'uuid': 'ENSG00000147889',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000147889.18',\n",
       "    'HGNC': 'HGNC:1787',\n",
       "    'end': 21995301,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'EZH2',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'EZH2',\n",
       "    'start': 148807257,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '7',\n",
       "    'uuid': 'ENSG00000106462',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000106462.12',\n",
       "    'HGNC': 'HGNC:3527',\n",
       "    'end': 148884321,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'BCL2',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'BCL2',\n",
       "    'start': 63123346,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '18',\n",
       "    'uuid': 'ENSG00000171791',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000171791.14',\n",
       "    'HGNC': 'HGNC:990',\n",
       "    'end': 63320128,\n",
       "    'tag': ['overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'NF1',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'NF1',\n",
       "    'start': 31094927,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '17',\n",
       "    'uuid': 'ENSG00000196712',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000196712.20',\n",
       "    'HGNC': 'HGNC:7765',\n",
       "    'end': 31382116,\n",
       "    'tag': ['overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'NSD1',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'NSD1',\n",
       "    'start': 177131830,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '5',\n",
       "    'uuid': 'ENSG00000165671',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000165671.22',\n",
       "    'HGNC': 'HGNC:14234',\n",
       "    'end': 177300213,\n",
       "    'tag': ['overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'PHF6',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'PHF6',\n",
       "    'start': 134373288,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': 'X',\n",
       "    'uuid': 'ENSG00000156531',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000156531.18',\n",
       "    'HGNC': 'HGNC:18145',\n",
       "    'end': 134428791,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'SAMHD1',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'SAMHD1',\n",
       "    'start': 36890229,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '20',\n",
       "    'uuid': 'ENSG00000101347',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000101347.11',\n",
       "    'HGNC': 'HGNC:15925',\n",
       "    'end': 36951893,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'SRSF2',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'SRSF2',\n",
       "    'start': 76734115,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '17',\n",
       "    'uuid': 'ENSG00000161547',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000161547.17',\n",
       "    'HGNC': 'HGNC:10783',\n",
       "    'end': 76737333,\n",
       "    'tag': ['ncRNA_host']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'CHD4',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'CHD4',\n",
       "    'start': 6570082,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '12',\n",
       "    'uuid': 'ENSG00000111642',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000111642.16',\n",
       "    'HGNC': 'HGNC:1919',\n",
       "    'end': 6614524,\n",
       "    'tag': ['ncRNA_host', 'overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'CHRNB4',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'CHRNB4',\n",
       "    'start': 78624111,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '15',\n",
       "    'uuid': 'ENSG00000117971',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000117971.12',\n",
       "    'HGNC': 'HGNC:1964',\n",
       "    'end': 78727754,\n",
       "    'tag': ['overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'CMPK1',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'CMPK1',\n",
       "    'start': 47333790,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '1',\n",
       "    'uuid': 'ENSG00000162368',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000162368.15',\n",
       "    'HGNC': 'HGNC:18170',\n",
       "    'end': 47394866,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'CYCS',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'CYCS',\n",
       "    'start': 25118656,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '7',\n",
       "    'uuid': 'ENSG00000172115',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000172115.9',\n",
       "    'HGNC': 'HGNC:19986',\n",
       "    'end': 25125260,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'ERCC1',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'ERCC1',\n",
       "    'start': 45407334,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '19',\n",
       "    'uuid': 'ENSG00000012061',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000012061.17',\n",
       "    'HGNC': 'HGNC:3433',\n",
       "    'end': 45478828,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'ESCO2',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'ESCO2',\n",
       "    'start': 27771949,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '8',\n",
       "    'uuid': 'ENSG00000171320',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000171320.15',\n",
       "    'HGNC': 'HGNC:27230',\n",
       "    'end': 27812640,\n",
       "    'tag': ['ncRNA_host']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'FGFR4',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'FGFR4',\n",
       "    'start': 177086905,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '5',\n",
       "    'uuid': 'ENSG00000160867',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000160867.15',\n",
       "    'HGNC': 'HGNC:3691',\n",
       "    'end': 177098144,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'FH',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'FH',\n",
       "    'start': 241497511,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '1',\n",
       "    'uuid': 'ENSG00000091483',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000091483.8',\n",
       "    'HGNC': 'HGNC:3700',\n",
       "    'end': 241519799,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'FOXA1',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'FOXA1',\n",
       "    'start': 37589552,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '14',\n",
       "    'uuid': 'ENSG00000129514',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000129514.8',\n",
       "    'HGNC': 'HGNC:5021',\n",
       "    'end': 37596059,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'HNRNPL',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'HNRNPL',\n",
       "    'start': 38836370,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '19',\n",
       "    'uuid': 'ENSG00000104824',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000104824.18',\n",
       "    'HGNC': 'HGNC:5045',\n",
       "    'end': 38852347,\n",
       "    'tag': ['overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'IDH2',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'IDH2',\n",
       "    'start': 90083045,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '15',\n",
       "    'uuid': 'ENSG00000182054',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000182054.10',\n",
       "    'HGNC': 'HGNC:5383',\n",
       "    'end': 90102477,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'MAP2K4',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'MAP2K4',\n",
       "    'start': 12020829,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '17',\n",
       "    'uuid': 'ENSG00000065559',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000065559.15',\n",
       "    'HGNC': 'HGNC:6844',\n",
       "    'end': 12143830,\n",
       "    'tag': ['ncRNA_host', 'overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'MAP2K7',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'MAP2K7',\n",
       "    'start': 7903843,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '19',\n",
       "    'uuid': 'ENSG00000076984',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000076984.18',\n",
       "    'HGNC': 'HGNC:6847',\n",
       "    'end': 7914478,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'MAP3K1',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'MAP3K1',\n",
       "    'start': 56815549,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '5',\n",
       "    'uuid': 'ENSG00000095015',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000095015.6',\n",
       "    'HGNC': 'HGNC:6848',\n",
       "    'end': 56896152,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'miRNA',\n",
       "    'symbol': 'MIR136',\n",
       "    'level': 'automatically annotated loci',\n",
       "    'displayName': 'MIR136',\n",
       "    'start': 100884702,\n",
       "    'source': 'ENSEMBL',\n",
       "    'chr': '14',\n",
       "    'uuid': 'ENSG00000207942',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000207942.2',\n",
       "    'HGNC': 'HGNC:31522',\n",
       "    'end': 100884783,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'miRNA',\n",
       "    'symbol': 'MIR146A',\n",
       "    'level': 'automatically annotated loci',\n",
       "    'displayName': 'MIR146A',\n",
       "    'start': 160485352,\n",
       "    'source': 'ENSEMBL',\n",
       "    'chr': '5',\n",
       "    'uuid': 'ENSG00000283733',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000283733.1',\n",
       "    'HGNC': 'HGNC:31533',\n",
       "    'end': 160485450,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'miRNA',\n",
       "    'symbol': 'MIR193A',\n",
       "    'level': 'automatically annotated loci',\n",
       "    'displayName': 'MIR193A',\n",
       "    'start': 31559996,\n",
       "    'source': 'ENSEMBL',\n",
       "    'chr': '17',\n",
       "    'uuid': 'ENSG00000207614',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000207614.1',\n",
       "    'HGNC': 'HGNC:31563',\n",
       "    'end': 31560083,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'miRNA',\n",
       "    'symbol': 'MIR222',\n",
       "    'level': 'automatically annotated loci',\n",
       "    'displayName': 'MIR222',\n",
       "    'start': 45747015,\n",
       "    'source': 'ENSEMBL',\n",
       "    'chr': 'X',\n",
       "    'uuid': 'ENSG00000207725',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000207725.3',\n",
       "    'HGNC': 'HGNC:31602',\n",
       "    'end': 45747124,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'miRNA',\n",
       "    'symbol': 'MIR9-3',\n",
       "    'level': 'automatically annotated loci',\n",
       "    'displayName': 'MIR9-3',\n",
       "    'start': 89368017,\n",
       "    'source': 'ENSEMBL',\n",
       "    'chr': '15',\n",
       "    'uuid': 'ENSG00000284329',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000284329.1',\n",
       "    'HGNC': 'HGNC:31646',\n",
       "    'end': 89368106,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'MTHFR',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'MTHFR',\n",
       "    'start': 11785723,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '1',\n",
       "    'uuid': 'ENSG00000177000',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000177000.13',\n",
       "    'HGNC': 'HGNC:7436',\n",
       "    'end': 11806455,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'POLE',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'POLE',\n",
       "    'start': 132623753,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '12',\n",
       "    'uuid': 'ENSG00000177084',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000177084.19',\n",
       "    'HGNC': 'HGNC:9177',\n",
       "    'end': 132687376,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'PRDM1',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'PRDM1',\n",
       "    'start': 105993463,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '6',\n",
       "    'uuid': 'ENSG00000057657',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000057657.17',\n",
       "    'HGNC': 'HGNC:9346',\n",
       "    'end': 106109939,\n",
       "    'tag': ['overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'PTGS2',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'PTGS2',\n",
       "    'start': 186671791,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '1',\n",
       "    'uuid': 'ENSG00000073756',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000073756.13',\n",
       "    'HGNC': 'HGNC:9605',\n",
       "    'end': 186680922,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'PTPN13',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'PTPN13',\n",
       "    'start': 86594315,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '4',\n",
       "    'uuid': 'ENSG00000163629',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000163629.13',\n",
       "    'HGNC': 'HGNC:9646',\n",
       "    'end': 86815171,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'RICTOR',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'RICTOR',\n",
       "    'start': 38937920,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '5',\n",
       "    'uuid': 'ENSG00000164327',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000164327.14',\n",
       "    'HGNC': 'HGNC:28611',\n",
       "    'end': 39074402,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'RTEL1',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'RTEL1',\n",
       "    'start': 63657810,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '20',\n",
       "    'uuid': 'ENSG00000258366',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000258366.12',\n",
       "    'HGNC': 'HGNC:15888',\n",
       "    'end': 63696253,\n",
       "    'tag': ['overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'SDHC',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'SDHC',\n",
       "    'start': 161314381,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '1',\n",
       "    'uuid': 'ENSG00000143252',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000143252.16',\n",
       "    'HGNC': 'HGNC:10682',\n",
       "    'end': 161363206,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'SETD2',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'SETD2',\n",
       "    'start': 47016428,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '3',\n",
       "    'uuid': 'ENSG00000181555',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000181555.22',\n",
       "    'HGNC': 'HGNC:18420',\n",
       "    'end': 47164113,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'SETDB1',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'SETDB1',\n",
       "    'start': 150926263,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '1',\n",
       "    'uuid': 'ENSG00000143379',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000143379.13',\n",
       "    'HGNC': 'HGNC:10761',\n",
       "    'end': 150964744,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'SLC3A2',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'SLC3A2',\n",
       "    'start': 62856004,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '11',\n",
       "    'uuid': 'ENSG00000168003',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000168003.19',\n",
       "    'HGNC': 'HGNC:11026',\n",
       "    'end': 62888880,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'SOX9',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'SOX9',\n",
       "    'start': 72121020,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '17',\n",
       "    'uuid': 'ENSG00000125398',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000125398.8',\n",
       "    'HGNC': 'HGNC:11204',\n",
       "    'end': 72126416,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'STIM1',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'STIM1',\n",
       "    'start': 3854527,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '11',\n",
       "    'uuid': 'ENSG00000167323',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000167323.13',\n",
       "    'HGNC': 'HGNC:11386',\n",
       "    'end': 4093210,\n",
       "    'tag': ['ncRNA_host', 'overlapping_locus']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'TNFRSF1A',\n",
       "    'level': 'verified loci',\n",
       "    'displayName': 'TNFRSF1A',\n",
       "    'start': 6328757,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '12',\n",
       "    'uuid': 'ENSG00000067182',\n",
       "    'strand': -1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000067182.9',\n",
       "    'HGNC': 'HGNC:11916',\n",
       "    'end': 6342114,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'IL13',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'IL13',\n",
       "    'start': 132656263,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '5',\n",
       "    'uuid': 'ENSG00000169194',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000169194.10',\n",
       "    'HGNC': 'HGNC:5973',\n",
       "    'end': 132661110,\n",
       "    'tag': ['']},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'symbol': 'TLR2',\n",
       "    'level': 'manually annotated loci',\n",
       "    'displayName': 'TLR2',\n",
       "    'start': 153684070,\n",
       "    'source': 'HAVANA',\n",
       "    'chr': '4',\n",
       "    'uuid': 'ENSG00000137462',\n",
       "    'strand': 1,\n",
       "    'ensembl_gene_id_version': 'ENSG00000137462.9',\n",
       "    'HGNC': 'HGNC:11848',\n",
       "    'end': 153706260,\n",
       "    'tag': ['']}]}]"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "935fc26f-ae36-48ac-bdf0-a033d76dc32e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1, 2, 3, 4]"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[1, 2] + [3,4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7d89715-b837-4bab-abaa-ed8ee97d9f2f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
